src/peer.rs

```diff
@@ -26,85 +22,38 @@ use crate::{Address, Error, Packet};
 ///
 /// ENet allows the association of arbitrary data with each peer.
 /// The type of this associated data is chosen through `T`.
-#[derive(Clone, Debug)]
-pub struct Peer<'a, T: 'a> {
-    inner: *mut ENetPeer,
-
-    _data: PhantomData<&'a mut T>,
-}
-
-/// A packet received directly from a `Peer`.
-///
-/// Contains the received packet as well as the channel on which it was received.
-#[derive(Debug)]
-pub struct PeerPacket<'b, 'a, T: 'a> {
-    /// The packet that was received.
-    pub packet: Packet,
-    /// The channel on which the packet was received.
-    pub channel_id: u8,
-
-    _priv_guard: PhantomData<&'b Peer<'a, T>>,
-}
-
-/// Describes the state a `Peer` is in.
-///
-/// The states should be self-explanatory, ENet doesn't explain them more either.
-#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
-#[allow(missing_docs)]
-pub enum PeerState {
-    Disconnected,
-    Connected,
-    Connecting,
-    AcknowledgingConnect,
-    ConnectionPending,
-    ConnectionSucceeded,
-    DisconnectLater,
-    Disconnecting,
-    AcknowledgingDisconnect,
-    Zombie,
+#[repr(transparent)]
+pub struct Peer<T> {
+    inner: ENetPeer,
+    _data: PhantomData<T>,
 }
 
-impl PeerState {
-    fn from_sys_state(enet_sys_state: _ENetPeerState) -> PeerState {
-        #[allow(non_upper_case_globals)]
-        match enet_sys_state {
-            _ENetPeerState_ENET_PEER_STATE_DISCONNECTED => PeerState::Disconnected,
-            _ENetPeerState_ENET_PEER_STATE_CONNECTING => PeerState::Connecting,
-            _ENetPeerState_ENET_PEER_STATE_ACKNOWLEDGING_CONNECT => PeerState::AcknowledgingConnect,
-            _ENetPeerState_ENET_PEER_STATE_CONNECTION_PENDING => PeerState::ConnectionPending,
-            _ENetPeerState_ENET_PEER_STATE_CONNECTION_SUCCEEDED => PeerState::ConnectionSucceeded,
-            _ENetPeerState_ENET_PEER_STATE_CONNECTED => PeerState::Connected,
-            _ENetPeerState_ENET_PEER_STATE_DISCONNECT_LATER => PeerState::DisconnectLater,
-            _ENetPeerState_ENET_PEER_STATE_DISCONNECTING => PeerState::Disconnecting,
-            _ENetPeerState_ENET_PEER_STATE_ACKNOWLEDGING_DISCONNECT => PeerState::AcknowledgingDisconnect,
-            _ENetPeerState_ENET_PEER_STATE_ZOMBIE => PeerState::Zombie,
-            val => panic!("unexpected peer state: {}", val),
-        }
+impl<'a, T> Peer<T>
+where
+    T: 'a,
+{
+    pub(crate) fn new(inner: &'a ENetPeer) -> &'a Peer<T> {
+        unsafe { &*(inner as *const _ as *const Peer<T>) }
```

Comment:
Oof, I'm not sure I like this. This feels quite hacky to me, but I see why you are doing it. The alternative would be having both a `Peer` and a `PeerMut` struct, right? That's definitely not nice either. This seems to be an advantage of the current design, where a `Peer` is owned but has a lifetime, so it can't just run out of scope too late.